2/contrib/s2i/run

```diff
@@ -181,6 +181,19 @@ new_password_hash=`obfuscate_password ${JENKINS_PASSWORD:-password} $old_salt`
 mkdir ${JENKINS_HOME}/logs
 ln -sf ${JENKINS_HOME}/logs /var/log/jenkins
 
+if [ "$(ls /usr/lib/jenkins/*.hpi 2>/dev/null)" ]; then
+   
+   for FILENAME in /usr/lib/jenkins/*.hpi ; do
+        basefilename=`basename $FILENAME .hpi`
+        otherfile="${JENKINS_HOME}/plugins/${basefilename}.*"
+        if [ "$(ls ${otherfile} 2>/dev/null)" ]; then
+          continue
+        fi
+	echo "Copying ${FILENAME} RPM installed Jenkins plugins to ${JENKINS_HOME} ..."  
+        cp --remove-destination $FILENAME ${JENKINS_HOME}/plugins/${basefilename}.jpi
+   done
+fi
```

Comment:
I wonder if we should just always copy, period.  Otherwise you risk not updating a dependent plugin to the necessary version if the version changed in the new image.

Reply:
The obvious downside would be users have no way to remove a plugin, or update it to a newer version than what's in the image.  (the former doesn't bother me, but the latter certainly does)

Reply:
Yes the later is what I tried to convey in https://github.com/openshift/jenkins/pull/425#issuecomment-348742162

The update of an existing dependency to a new version will work as things are currently constructed because we have links at the moment from /var/lib/jenkins/plugins to /usr/lib/jenkins.

But on subsequent upgrades that will break down, as we've started copying.  The next time a dependencies version needs to increase, it would no longer be correct.  So linking instead of copying helps in this case.  The "master", "original" version is in /usr/lib/jenkins.

So, yes, implies moving back to links from /var/lib/jenkins/plugins to /usr/lib/jenkins ...

Now, if the user wants to override, they go through the various paths to install-plugins.sh, then they break the links, but they have chose to manage things on their own and that point.

And all this just made me realize that we need to update install-plugins.sh and its use of `cp` to include the use of `--remove-destination` to probably override a link .... 

... sign, this is never going to end, is it.

Reply:
with symlinks, presumably if they update a plugin version (Via the console), that will update the .jpi file in /usr/lib (where the symlink points).  Since that's not a PV location, the next time jenkins restarted, they'd lose the update anyway.  (this is a problem in the existing implementation).

Basically I think the only safe thing to do is copy the plugins, and only do it the first time a new version of the image is run against a given /var/lib/jenkins dir.  Or do a timestamp check and don't overwrite files that are newer than what's in the image.

Reply:
I guess timestamps won't work since:

image1 contains f1.a at t1
image2 is shipped with f1.c at t2
user updates f1.a to f1.b at t3
user picks up image2, but f1.c won't overwrite f1.b because it's not "newer"

Reply:
No, updates via the console will update the .jpi file in /var/lib/jenkins/plugins ... it will break the link to /usr/lib.  The jenkins home that jenkins works off of is set to the PV.

If we could somehow determine that we are in an upgrade scenario (I'm guessing your timestamp exploration was an attempt at that), then yeah, we could process things differently.

Reply:
Confirmed via manual test that it breaks the link when upgrading a plugin via the jenkins console

Reply:
A back up file exists in fact that still has the link ... so if you were to revert, things would work as expected.

Reply:
ok so where does this leave us? Preferring to create links in the PV, but only if there's no non-link file already in the PV?

Reply:
yes, i believe that is what we should do .... I'll get started on the rework